### DISCUSSION TITLE
Variable Aware Kitty Session

### DISCUSSION THREAD

**Discussion by ghost** [kovidgoyal/kitty#5917] [Q&A] [answered]
I am trying to run kitty via `kitty --session example.conf` with a pre-chosen variable that determines the kitty window behavior, i.e. I pick what program to run in a script and `example.conf` can reference that variable. 

I would also like to have remote_control check the status of the variable, and run certain kitty commands. Is this possible without using an external package? Hopefully this would make it possible to close Kitty when an external program closes. 

I do a similar thing with a file watching package (`watchman`) to monitor files and modify Kitty behavior based on file changes. However, I think in this case the problem shouldn't require another program, IMO. 

Reading and changing Environment variables (i.e. `pam_env`) is quite different on different operating systems; though I hope that a solution could be somewhat OS-agnostic.

**Comment by kovidgoyal** [comment]
I dont follow what you want to do. Elaborate.

**Reply by ghost** [reply]
Say I want to move to a specific folder, and I choose that folder in a script prior to running `kitty --session example.conf`, I want the Kitty window to open in that directory.

Ideally I could also monitor a variable, say $ANOTHER_FOLDER, and automatically open a kitty tab with that folder open if is created (the folder would be created by an external program, not done manually by me because then there would be no reason for automation). 

I would prefer to not run a script _after_ the kitty window is launched for a variety of reasons.

**Comment by kovidgoyal** [comment]
On Sat, Jan 21, 2023 at 09:32:54PM -0800, Carson Surmeier wrote:
> Say I want to change a specific folder, and I choose that folder in a script prior to running `kitty --session example.conf` and the Kitty window opens in that directory.

kitty --directory $whatever --session whatever

or

kitty --session <(echo "cd $whatever\n$(cat example.conf)")

> 
> Ideally I could also monitor a variable, say $ANOTHER_FOLDER, and open a kitty tab with that folder open if is created (would be created by an external program). 

Use remote control for that.

> 
> I would prefer to not run a script _after_ the kitty window is launched for a variety of reasons. 

You want to run a script to change the directory of an existing shell
running inside an existing kitty window? Use remote control with
send-text for that. kitty @ send-text --match state:self "cd $whatever\r"

Change --match to match whatever kitty window you want.

**Reply by ghost** [reply]
The directory example is just an example, a bad one at that. Apologies. I am aware of the `--directory` flag, I am talking about being able to use variables in the `.conf` file specified in the kitty command. 

As for 2), I am aware of kitty's remote control capabilities and will write some python to solve the example I gave. 

As for 3), again, using `cd $FOLDER` was just an example but thank you for the solution using send-text.

**Comment by kovidgoyal** [comment, marked as answer]
On Sat, Jan 21, 2023 at 09:53:04PM -0800, Carson Surmeier wrote:
> The directory example is just an example, a bad one at that. Apologies. I am aware of the `--directory` flag, I am talking about being able to use variables in the `.conf` file specified in the kitty command. 

No, there is no facility for using variables in such files. If you need
to, then dynamically generate them from a template, which is trivial
to do in most shells/scripting languages. I showed you an example
before of dynamically generating the session file.

I suppose one could add support for evaluating env vars in the session
file as well, the main issue is really the syntax for said env vars,
since a session file can contain launch commands which themselves can
run shell commands containing env vars one cant re-use shell variable
syntax. So one would need to come up with some syntax not likely to
conflict with other things one is likely to run in launch commands. Or
maybe only eval env vars outside of the args to a launch command.
